nautobot/dcim/views.py

```diff
@@ -3887,151 +3887,97 @@ def extra_context(self):
+        context = super().get_extra_context(request, instance)
 
-        vc_form = forms.VirtualChassisForm(request.POST, instance=virtual_chassis)
-        vc_form.fields["master"].queryset = members_queryset
-        formset = VCMemberFormSet(request.POST, queryset=members_queryset)
+        if self.action == "update":
+            VCMemberFormSet = modelformset_factory(
+                model=Device,
+                form=forms.DeviceVCMembershipForm,
+                formset=forms.BaseVCMemberFormSet,
+                extra=0,
+            )
+            members_queryset = instance.members.select_related("rack").order_by("vc_position")
 
-        if vc_form.is_valid() and formset.is_valid():
-            with transaction.atomic():
-                # Save the VirtualChassis
-                vc_form.save()
+            if request.method == "POST":
+                formset = VCMemberFormSet(request.POST, queryset=members_queryset)
+            else:
+                formset = VCMemberFormSet(queryset=members_queryset)
 
-                # Nullify the vc_position of each member first to allow reordering without raising an IntegrityError on
-                # duplicate positions. Then save each member instance.
-                members = formset.save(commit=False)
-                devices = Device.objects.filter(pk__in=[m.pk for m in members])
-                for device in devices:
-                    device.vc_position = None
-                    device.save()
-                for member in members:
-                    member.save()
+            context.update(
+                {
+                    "formset": formset,
+                    "vc_form": self.get_form_class()(instance=instance),
+                    "return_url": self.get_return_url(request, instance),
+                }
+            )
 
-            return redirect(virtual_chassis.get_absolute_url())
+        elif self.action == "retrieve":
+            members = Device.objects.restrict(request.user).filter(virtual_chassis=instance)
+            context.update({"members": members})
 
-        return render(
-            request,
-            "dcim/virtualchassis_edit.html",
-            {
-                "vc_form": vc_form,
-                "formset": formset,
-                "return_url": self.get_return_url(request, virtual_chassis),
-            },
-        )
+        return context
 
+    def form_save(self, form, **kwargs):
+        obj = super().form_save(form, **kwargs)
 
-class VirtualChassisDeleteView(generic.ObjectDeleteView):
-    queryset = VirtualChassis.objects.all()
+        if self.action == "update":
+            context = self.get_extra_context(self.request, obj)
+            formset = context.get("formset")
 
+            if formset.is_valid():
+                with transaction.atomic():
+                    members = formset.save(commit=False)
+                    Device.objects.filter(pk__in=[m.pk for m in members]).update(vc_position=None)
```

Comment:
Can you confirm that the Device object-change records still get correctly populated when using `.objects.update()` instead of the previous for-loop implementation?